lib/CL/devices/level0/level0-driver.cc

```diff
@@ -0,0 +1,2239 @@
+  Res = zeDeviceGetImageProperties(DeviceHandle, &ImageProperties);
+  if (Res != ZE_RESULT_SUCCESS)
+    return;
+
+  ze_device_external_memory_properties_t ExternalMemProperties{};
+  ExternalMemProperties.stype =
+      ZE_STRUCTURE_TYPE_DEVICE_EXTERNAL_MEMORY_PROPERTIES;
+  ExternalMemProperties.pNext = nullptr;
+  Res =
+      zeDeviceGetExternalMemoryProperties(DeviceHandle, &ExternalMemProperties);
+  if (Res != ZE_RESULT_SUCCESS)
+    return;
+
+  POCL_MSG_PRINT_LEVEL0("all device info collected\n");
+
+  /**************************************************************************/
+
+  // TODO
+  ClDev->profiling_timer_resolution = 1;
+
+  // Fixed props
+  {
+    ClDev->compiler_available = CL_TRUE;
+    ClDev->linker_available = CL_TRUE;
+    ClDev->has_own_timer = CL_TRUE;
+    ClDev->disable_pocl_opencl_headers = CL_TRUE;
+
+    ClDev->preferred_vector_width_char = 16;
+    ClDev->preferred_vector_width_short = 8;
+    ClDev->preferred_vector_width_int = 4;
+    ClDev->preferred_vector_width_long = 1;
+    ClDev->preferred_vector_width_float = 1;
+    ClDev->preferred_vector_width_double = 1;
+    ClDev->preferred_vector_width_half = 8;
+    ClDev->native_vector_width_char = 16;
+    ClDev->native_vector_width_short = 8;
+    ClDev->native_vector_width_int = 4;
+    ClDev->native_vector_width_long = 1;
+    ClDev->native_vector_width_float = 1;
+    ClDev->native_vector_width_double = 1;
+    ClDev->native_vector_width_half = 8;
+
+    ClDev->has_64bit_long = CL_TRUE;
+
+    ClDev->endian_little = CL_TRUE;
+    ClDev->parent_device = NULL;
+    ClDev->max_sub_devices = 0;
+    ClDev->num_partition_properties = 0;
+    ClDev->partition_properties = NULL;
+    ClDev->num_partition_types = 0;
+    ClDev->partition_type = NULL;
+    ClDev->max_constant_args = 8;
+    ClDev->host_unified_memory = Integrated ? CL_TRUE : CL_FALSE;
+    ClDev->min_data_type_align_size = MAX_EXTENDED_ALIGNMENT;
+    ClDev->global_var_max_size = 64 * 1024;
+
+    ClDev->execution_capabilities = CL_EXEC_KERNEL;
+    ClDev->address_bits = 64;
+    ClDev->vendor = "Intel Corporation";
+    ClDev->vendor_id = 0x8086;
+    ClDev->profile = "FULL_PROFILE";
+
+    ClDev->num_serialize_entries = 1;
+    ClDev->serialize_entries = LEVEL0_SERIALIZE_ENTRIES;
+    ClDev->llvm_cpu = nullptr;
+    ClDev->llvm_target_triplet = "spir64-unknown-unknown";
+    ClDev->generic_as_support = CL_TRUE;
+    ClDev->supported_spir_v_versions = "SPIR-V_1.2";
+    ClDev->on_host_queue_props = CL_QUEUE_PROFILING_ENABLE;
+    ClDev->version_of_latest_passed_cts = HOST_DEVICE_LATEST_CTS_PASS;
+  }
+
+  /**************************************************************************/
+
+  // deviceProperties
+  {
+    switch (DeviceProperties.type) {
+    case ZE_DEVICE_TYPE_CPU:
+      ClDev->type = CL_DEVICE_TYPE_CPU;
+      break;
+    case ZE_DEVICE_TYPE_GPU:
+      ClDev->type = CL_DEVICE_TYPE_GPU;
+      break;
+    default:
+      ClDev->type = CL_DEVICE_TYPE_CUSTOM;
```

Comment:
Should we just abort here since if we do not know what do with the level0 custom devices?

Reply:
I'd prefer to not abort. I can replace the break with return; that leaves the device in existing but not available state.

Reply:
fixed